### PR TITLE
Command discovery/description

### PR DESCRIPTION
**Pull request by kythyria**
This is quite draft; I'm not quite sure some of the syntax is right. It tries to address #112 without requiring that the banner sent by servers be EVEN LONGER.

There's no cap, because `ISUPPORT` is used instead.

### CODE REVIEW

**Comment by SaberUK** [comment]
:+1:

File: extensions/command-discovery-3.3.md [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+The last three are of somewhat special nature: `text` must be the last parameter
+other than `label`, since spaces are permitted in this--the payload of, say,
+`PRIVMSG` is a `text` parameter--and `label` does not define a parameter at all,
+but rather suggests a name to use when displaying the command in a GUI, such as
+if it is presented as a menu item.
+
+## Example
+Suppose `RPL_CMDLIST` and `RPL_CMDLISTEND` are `297` and `299` respectively.
+```
+S: 005 EXCEPTS INVEX CMDLIST=571
```
Comment (SaberUK) [inline extensions/command-discovery-3.3.md]: There's probably no point in having a value for the CMDLIST token. If a client supports it they will know what numeric it uses anyway.

**Comment by robotoad** [comment]
How would this information be localized?

**Comment by kaniini** [comment]
@robotoad brings up a good point actually, not everyone speaks english, and not everyone on a predominantly spanish language server for example, wants their commands in spanish etc.

**Comment by SaberUK** [comment]
I was working on a proposal a while back for a language cap. If anyone is interested I can finish that and submit it.

**Comment by grawity** [comment]
I wonder if the server should provide a command ID (to let clients override the label or just recognize it) as well as a few primary translations (including an `en` translation is strongly recommended).

(Or, gettext-style, just English + translations.)

**Comment by kaniini** [comment]
Adding type and context as separate things seems unintuitive to me.  Why can't context *be* type?

**Comment by kythyria** [comment]
> Adding type and context as separate things seems unintuitive to me. Why can't context be type?

@kaniini Because context applies to the entire command, type applies to individual parameters, and some of the values for each don't make sense for the other (eg, `sc` doesn't make sense as a context, and `self` doesn't as a type).

Consider `KICK`. It has a context of `nickchan`, and parameters of `channel`, `nick`, and `text`. `nickchan` doesn't make sense as a type, because why would you put both those things in one parameter. But it certainly works as a context, because `KICK` removes a specific user from a specific channel.

**Comment by kaniini** [comment]
@spb mentions that the format provided by this may not map well to some esoteric things such as charybdis' `SCAN` command.  this is probably not of much relevance however, since the odds of charybdis adopting 3.3 is nil.  but there may be other commands like it.

a better study, thinking about it, would be to consider how this would map to IRCv3.3 `XINFO`, which has similar properties to `SCAN`.

**Comment by spb** [comment]
More than that, this is entirely the wrong approach to implementing and documenting extensions to the protocol.

As the two examples above illustrate, *any* machine-readable description of command syntax that you can come up with will have shortcomings and somebody will want to implement a command that can't be described in that way. More than that, it can't describe to the client what the effect of that command will be.

In the general case, and I suspect in the majority of real-world cases, adding a sensible user interface for non-standard commands requires an understanding of what the command *does*, not just what its argument types are. Any sensible GUI for a particular command requires that the client actually understands the command, and hasn't just taken it out of a list of supported stuff.

To that end, it'd be far better to provide a means for a server to document which feature extensions it supports, which can then be documented either inside or outside the IRCv3 spec, and clients can then add support for those features as and when they want to. Trying to have the server define the client's UI will only end up in reinventing the web browser. That means might or might not look a lot like ISUPPORT.

**Comment by SaberUK** [comment]
>To that end, it'd be far better to provide a means for a server to document which feature extensions it supports

As previously discussed, every major server implementation allows users to define custom aliases. You can't write a specification for that.

**Comment by spb** [comment]
You also can't auto-generate a GUI for it.

**Comment by kythyria** [comment]
@SaberUK You can, sort of. It will look a lot like this one. Compare [XEP-0050](http://xmpp.org/extensions/xep-0050.html) and [XEP-0004](http://xmpp.org/extensions/xep-0004.html) which add up to an attempt to solve a similar problem. It works reasonably well, actually, since the latter's data model is quite simple and so doesn't demand much from a GUI.

**Comment by kythyria** [comment]
@spb
> You also can't auto-generate a GUI for it.

You can for some simple yet common cases, since very little GUI is needed, eg, for `CS KICK`. Channel and nick are implied by, eg, selecting the user from a list of users in the channel, and then prompting for the reason field is easy.

My intention was as much about doing optional parameters the way most clients do for `/part` as any sort of GUI.

**Comment by spb** [comment]
Is the intention to write an IRC specification that works "for some simple yet common cases"?

Because that'd be a pretty bad specification.

**Comment by kythyria** [comment]
There's not much complex you can fit in an IRC message, but if there were a ton of non-context fields in a command that nonetheless gets exposed in a GUI, you could do that by presenting a simple tabular layout, like XEP-0004 clients do. A client could also organise subcommand-heavy verbs, such as `CS`, using a submenu approach.

And this won't ever cover every nuance of what a client can send to the server. That would require a much bigger hunk of data that would be a pain to fit in-band, since you'd need to cover the values of enumerated fields, probably group subcommands explicitly, and I'm not even going to touch modes.

**Comment by M2Ys4U** [comment]
I agree with @spb.

Any auto-generated GUI will have an awful UX. You may be able to encode *syntactic* information but you'll be missing *semantic* information.

----

I still fail to see any compelling rationale for this.

What server vendors should be doing is *specifying* any non-standard commands. If they're any good they'll probably be adopted by other vendors (perhaps even published by the IRCv3 WG, who knows!). Publishing a spec also helps to reduce namespace collisions.

When there's a specification, clients can be notified by including an `RPL_ISUPPORT` token for the spec and toggle any helpful UI.

**Comment by SaberUK** [comment]
@M2Ys4U I recommend reading the entire comment thread for information why that is not feasible.

**Comment by kaniini** [comment]
I think before we really move on this, we need to see something in action as a PoC so we can weigh how useful it really is.  I plan to submit an RFC which requires that large changes be accompanied with a PoC.

**Comment by M2Ys4U** [comment]
@SaberUK I have, I don't see why it's not feasible.

**Comment by kythyria** [comment]
@M2Ys4U You seriously think it's feasible for every single server developer, server *module* developer, and network adminstrator to a) write a proper spec of the extensions introduced, and b) convince the developers of a large number of clients to implement something so niche?

**Comment by kaniini** [comment]
@kythyria Right now the proposed example is usually some ZNC detach channel thing.  In that case I'm not convinced, why can't ZNC make a specification for detaching and reattaching channels instead?  (In fact, a high quality specification documenting bouncer-specific extensions to IRC would be a wonderful contribution to IRCv3.)  The second big example is `NS` and `NICKSERV` aliases, both of which are then proposed as a hint to `IDENTIFY`, which is deprecated by SASL for over 4 years now, so that's kind of pointless to document.

This is why I say a PoC is necessary -- we need to get into a realistic mindset about this functionality in order to properly evaluate it.  Otherwise it becomes a "wouldn't it be nice if..." scenario.  Lots of things are nice, but less things are practical, and spending time on impracticalities slows the IRCv3 process.

**Comment by kythyria** [comment]
@kaniini The other thing is `CHANSERV`. Part of the reason using it is obnoxious compared to `MODE` and `KICK` and so forth is because you *have* to manually specify the channel each time, as that context doesn't travel over the wire.

**Comment by M2Ys4U** [comment]
@kythyria: It's not like I'm asking developers to write an Internet Draft and submit it to the IETF for publication as an RFC on the Standards Track.

It'll be enough to publish a document stating:

1. How the command works (the syntax),
2. What the command does, and
3. Why the command exists.

That should be enough information for any client that wants to implement a UI for it to do so in a way that makes sense. This current spec only deals with `1`

If something is "so niche" that nobody would write a spec for it then surely it's "so niche" that clients don't need a badly integrated UI for it.

Besides, isn't the entire point of the IRCv3 WG to specify, come to a consensus on, and standardise extensions to the IRC protocol?

**Comment by SaberUK** [comment]
@M2Ys4U Here is a specification for arbitrary server aliases as supported by every major server implementation:

Servers may implement any command. Any number and type of arguments are possible.

**Comment by kythyria** [comment]
@SaberUK @M2Ys4U We can extend that:

1. Servers may implement any command. Any number and type of arguments are possible.
2. These commands do *something*. This may or may not correspond to what the name implies, and may or may not happen when you try to use them.
3. It exists because someone thought it was a good idea. Or was drunk. Or a furry.

(I'm guilty of that last one, although because this spec doesn't exist it's triggered by `CTCP-ACTION` starting with a particular string)

**Comment by clokep** [comment]
I still don't see what specific problem this is attempting to solve. This solution seems to be conflating a whole bunch of unrelated issues, as @M2Ys4U alludes to. Additionally, at best you'll end up with a confusing UI that talks in terms of protocol level stuff instead of things the user actually cares about.

In terms of comments for the actual proposal: I strongly dislike the idea of having a label passed along with a command. The major issue I see with this is localization, which has been mentioned before. Even if multiple localized versions are provided, you'll likely run into the issue that IRC doesn't really have a defined character encoding, although most people seem to use UTF-8 now.

One of the arguments that keeps coming up is that it isn't possible since a server can implement arbitrary aliases. How does that behoove a server if a client can't interact without manual commands? Is an overly complicated spec like this really the answer? What aliases are pretty really declaring? Would it make more sense to declare a list of aliases and the "real" command they point to?

Regardless, it seems like some of the commands (e.g. NICKSERV) couldn't even be described in this format since the syntax of each sub-command that can be sent to NICKSERV is fairly arbitrary. If there's an ulterior motive of wanting *specific* functionality...we should concentrate on that (I think the idea of having a bouncer spec is a great idea!)

**Comment by robotoad** [comment]
Following are some issues from the perspective of client developers (please don't take my objections as anything more than a healthy part of the vetting process):

* Localization of user-visible text. @SaberUK has offered a solution for this.
* Native functionality only possible with knowledge of the meaning of a command. For example, typing "/msg " in an input box might display a popover of recently messaged nicknames. Automatic command discovery couldn't possibly supply enough contextual information for every possible interface.
* Poor user experience via auto-generated GUIs that couldn't possibly support all environments.
* Inability to display command help without an active connection to a server that has supplied the information. It could be cached, but it would need to be organized by server and would scale poorly for large lists in terms of usability.
* A move away from codifying extensions toward automation of arbitrary whim without intervention or knowledge on the part of client authors or the greater community, which might benefit from standardization of a useful feature.

A counterargument has been made that it isn't feasible for servers to document their additions or for client authors to track them. As I said before, if servers are to go through the trouble of supplying command syntax and localized help descriptions, it stands to reason the information could be placed in a published document since it's practically a specification already. If it's difficult to imagine client authors tracking published specifications, why would it be easier to imagine them implementing automated command parsing and GUI generation?

In the face of IRCv3's attempts at standardization, this feels like a reversal of direction by encouraging servers to go their own way. In my opinion, this proposal won't be treated as a bridge toward standardization; it will be a crutch used to avoid standardization.

**Comment by kaniini** [comment]
@robotoad that is basically my concern with it, we need to be very careful because we're playing with fire :)

**Comment by attilamolnar** [comment]
> In the face of IRCv3's attempts at standardization, this feels like a reversal of direction by encouraging servers to go their own way. In my opinion, this proposal won't be treated as a bridge toward standardization; it will be a crutch used to avoid standardization.

I agree. I support the idea of documenting some of the currently existing major features (and the commands that go with them) such as callerid, SA* commands and silence so clients can properly integrate them.
On the other hand, I don't think it's realistic to document every feature (command) because there are lots of oper-only features that opers should be expected to be able to use without any assistance from the client. For example the `LOADMODULE`, `NICKLOCK`, `MKPASSWD` commands are of no interest to most users. Some oper-only features could use a proper UI, such as displaying the output of a clone check but I don't think we should worry about things like that for now.

File: extensions/command-discovery-3.3.md [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+```
+sc          | Subcommand; literal text to include to invoke this one.
+chan        | Channel name
+nick        | Nickname
+user        | Services username (which is usually a nick)
+target      | Nick or channel, or possibly a targetlist.
+pattern     | Pattern such as a wildcarded hostmask
+other       | Some other parameter
+text        | Long text (must be the last non-label parameter!)
+label       | Not actually parameter. UI text to display for this command.
```
Comment (ProgVal) [inline extensions/command-discovery-3.3.md]: Could you add an integer type?

**Comment by ProgVal** [comment]
I think it would be nice to make it possible for bots (and services) to advertise their commands this way.
Do you think it should be in this spec, or in an other one once this PR is merged?

**Comment by DanielOaks** [comment]
I think this would be a very bad type of reply to standardise right now. As others have mentioned, I think this would hurt the standardisation of new IRC features and commands much more than help it. I think this feature directly opposes the working group's goals, and would be used to stifle the standardisation of client commands between different server software. Unless shown otherwise, I don't support this feature in any way.

**Comment by attilamolnar** [comment]
Closing due to lack of support. Moreover it was mentioned that this should be prototyped and demoed before moving forward which have not happened.

**Pull request closed without merging**